src/document_graph/document.cpp

```diff
@@ -0,0 +1,270 @@
+
+        id = h_itr->id;
+        creator = h_itr->creator;
+        created_date = h_itr->created_date;
+        certificates = h_itr->certificates;
+        content_groups = h_itr->content_groups;
+        hash = hashContents();
+
+        // this should never happen, only if hash algorithm somehow changed
+        eosio::check (hash == _hash, "fatal error: provided and indexed hash does not match newly generated hash");
+    }
+
+    void Document::emplace () 
+    {
+        require_auth(creator);
+        hash = hashContents();
+
+        document_table d_t(contract, contract.value);
+        auto hash_index = d_t.get_index<eosio::name("idhash")>();
+        auto h_itr = hash_index.find(hash);
+
+        // if this content exists already, error out and send back the hash of the existing document
+        eosio::check(h_itr == hash_index.end(), "document exists already: " + readableHash(hash));
+
+        d_t.emplace(contract, [&](auto &d) {
+            id = d_t.available_primary_key(); 
+            created_date = eosio::current_time_point();
+            d = *this;
+        });
+    }
+
+    Document Document::getOrNew (eosio::name _contract, eosio::name _creator, std::vector<ContentGroup> contentGroups) 
+    {
+        Document document {};
+        document.content_groups = contentGroups;
+        document.hash = document.hashContents();
+
+        Document::document_table d_t(_contract, _contract.value);
+        auto hash_index = d_t.get_index<eosio::name("idhash")>();
+        auto h_itr = hash_index.find(document.getHash());
+
+        // if this content exists already, return this one
+        if (h_itr != hash_index.end()) {
+            document.creator = h_itr->creator;
+            document.created_date = h_itr->created_date;
+            document.certificates = h_itr->certificates;
+            document.id = h_itr->id;
+            return document;
+        }
+
+        require_auth(_creator);
+        return Document (_contract, _creator, contentGroups);
+    }
+
+    Document Document::getOrNew (eosio::name contract, eosio::name creator, ContentGroup contentGroup)
+    {
+        return getOrNew (contract, creator, rollup (contentGroup));
+    }
+
+    Document Document::getOrNew (eosio::name contract, eosio::name creator, Content content) 
+    {
+        return getOrNew (contract, creator, rollup (content));
+    }
+
+    Document Document::getOrNew (eosio::name contract, eosio::name creator, const std::string &label, const Content::FlexValue &value)
+    {
+        return getOrNew (contract, creator, rollup (Content (label, value)));
+    }
+
+    void Document::setCreator(eosio::name &creator)
+    {
+        creator = creator;
+    }
+
+    eosio::name Document::getCreator()
+    {
+        return creator;
+    }
+
+    eosio::checksum256 Document::getHash()
+    {
+        return hash;
+    }
+
+    void Document::add(ContentGroup cg)
+    {
+        content_groups.push_back(cg);
+    }
```

Comment:
As a rule of thump I like to use when setting posible heavy objects is moving instead of copying if they are no longer required. As in this case the object paramater cg, will no longer be used after pushing it to the content group we could move it. 
`content_groups.push_back(cg);`

Reply:
I'm not sure this is needed anyways, but if so, I agree.  I will remove the method for now.